From 685c5322f81b957b314c970a72f43de70fbfc761 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Thu, 6 Jul 2023 16:53:36 +0530
Subject: [PATCH] ath12k: fix caldata download issue in qcn6432

Currently, we are passing default caldata file (ATH12K_DEFAULT_CAL_FILE)
for qcn6432 soc due to this ath12k_core_firmware_request function returns
ENOENT error while file downloaded according to userpd_id.

Fix this issue by passing caldata file according to userpd_id for each
ab instead to default.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/hw.h  | 2 ++
 drivers/net/wireless/ath/ath12k/qmi.c | 7 +++++++
 2 files changed, 9 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/hw.h
+++ b/drivers/net/wireless/ath/ath12k/hw.h
@@ -84,6 +84,8 @@
 #define ATH12K_BOARD_API2_FILE		"board-2.bin"
 #define ATH12K_DEFAULT_BOARD_FILE	"board.bin"
 #define ATH12K_DEFAULT_CAL_FILE		"caldata.bin"
+#define ATH12K_QMI_DEF_CAL_FILE_PREFIX  "caldata_"
+#define ATH12K_QMI_DEF_CAL_FILE_SUFFIX  ".bin"
 #define ATH12K_AMSS_FILE		"amss.bin"
 #define ATH12K_AMSS_DUALMAC_FILE	"amss_dualmac.bin"
 #define ATH12K_M3_FILE			"m3.bin"
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -4536,10 +4536,21 @@ static int ath12k_qmi_load_bdf_qmi(struc
 			snprintf(filename, sizeof(filename), "cal-%s-%s.bin",
 				 ath12k_bus_str(ab->hif.bus), dev_name(dev));
 			fw_entry = ath12k_core_firmware_request(ab, filename);
+
 			if (!IS_ERR(fw_entry))
 				goto success;
 
-			fw_entry = ath12k_core_firmware_request(ab, ATH12K_DEFAULT_CAL_FILE);
+			snprintf(filename, sizeof(filename), "%s", ATH12K_DEFAULT_CAL_FILE);
+
+			if (ab->userpd_id) {
+				snprintf(filename, sizeof(filename), "%s%d%s",
+					 ATH12K_QMI_DEF_CAL_FILE_PREFIX,
+					 ab->userpd_id,
+					 ATH12K_QMI_DEF_CAL_FILE_SUFFIX);
+			}
+
+			fw_entry = ath12k_core_firmware_request(ab, filename);
+
 			if (IS_ERR(fw_entry)) {
 				ret = PTR_ERR(fw_entry);
 				ath12k_warn(ab,
